### PR TITLE
Add Adocean bidder

### PR DESCRIPTION
**Pull request by AndriyPavlyuk**

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+                        .build();
+
+                final String bidCurrency = adoceanResponse.getCurrency() != null
+                        ? adoceanResponse.getCurrency()
+                        : DEFAULT_BID_CURRENCY;
+                final BidderBid bidderBid = BidderBid.of(updatedBid, BidType.banner, bidCurrency);
+                bidderBids.add(bidderBid);
+            }
+        }
+        return Result.of(bidderBids, errors);
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java:257]:
```

        final List<BidderError> errors = new ArrayList<>();
        final List<BidderBid> bidderBids = adoceanResponses.stream()
                .filter(adoceanResponse -> !adoceanResponse.getError().equals("true"))
                .filter(adoceanResponse -> auctionIds != null
                        && StringUtils.isNotBlank(auctionIds.get(adoceanResponse.getId())))
                .map(adoceanResponse -> BidderBid.of(createBid(auctionIds, adoceanResponse), BidType.banner,
                        getBidCurrency(adoceanResponse)))
                .collect(Collectors.toList());

        return Result.of(bidderBids, errors);
    }

    private static Bid createBid(Map<String, String> auctionIds, AdoceanResponseAdUnit adoceanResponse) {
        return Bid.builder()
                .id(adoceanResponse.getId())
                .impid(auctionIds.get(adoceanResponse.getId()))
                .adm(getAdm(adoceanResponse))
                .price(new BigDecimal(adoceanResponse.getPrice()))
                .w(Integer.valueOf(adoceanResponse.getWidth()))
                .h(Integer.valueOf(adoceanResponse.getHeight()))
                .crid(adoceanResponse.getCrid())
                .build();
    }

    private static String getBidCurrency(AdoceanResponseAdUnit adoceanResponse) {
        return adoceanResponse.getCurrency() != null
                ? adoceanResponse.getCurrency()
                : DEFAULT_BID_CURRENCY;
    }
```

File: src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+        // when
+        final Result<List<HttpRequest<Void>>> result = adoceanBidder.makeHttpRequests(bidRequest);
+
+        // then
+        assertThat(result.getErrors()).hasSize(1);
+        assertThat(result.getErrors().get(0).getMessage()).startsWith("Cannot deserialize instance");
+    }
+
+    @Test
+    public void makeHttpRequestsShouldSetReturnErrorIfRequestExists() {
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java:92]: `Set` is redundant

File: src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+                        .ext(ExtUser.builder()
+                                .consent("COwK6gaOwK6gaFmAAAENAPCAAAAAAAAAAAAAAAAAAAAA."
+                                        + "IFoEUQQgAIQwgIwQABAEAAAAOIAACAIAAAAQAIAgEAACEAAAAAgAQBAAAAAAAGBAAgAAAAAA"
+                                        + "AFAAECAAAgAAQARAEQAAAAAJAAIAAgAAAYQEAAAQmAgBC3ZAYzUw").build())
+                        .build())
+                .imp(asList(Imp.builder()
+                                .id("ao-test")
+                                .banner(Banner.builder().format(singletonList(Format.builder().w(300).h(250).build()))
+                                        .id("banner_id").build()).ext(mapper.valueToTree(ExtPrebid.of(null,
+                                ExtImpAdocean.of("myao.adocean.pl", "tmYF.DMl7ZBq.Nqt2Bq4FutQTJfTpxCOmtNPZoQUDcL.G7",
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java:105]: Does Banner and User is really necessary for this test ?

Also use codeStyle like this
```
   Imp.builder()
                                .id("ao-test")
                                .ext(mapper.valueToTree(ExtPrebid.of(null,
                                        ExtImpAdocean.of("myao.adocean.pl", "tmYF.DMl7ZBq.Nqt2Bq4FutQTJfTpxCOmtNPZoQUDcL.G7",
                                                "adoceanmyaozpniqis"))))
                                .build()))
```

File: src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+                        .ext(ExtUser.builder()
+                                .consent("COwK6gaOwK6gaFmAAAENAPCAAAAAAAAAAAAAAAAAAAAA.IFoEUQQgAIQwgIwQABAEAAAAOIAAC"
+                                        + "AIAAAAQAIAgEAACEAAAAAgAQBAAAAAAAGBAAgAAAAAAAFAAECAAAgAAQARAEQAAAAAJAAIAAgA"
+                                        + "AAYQEAAAQmAgBC3ZAYzUw").build())
+                        .build())
+                .imp(singletonList(Imp.builder()
+                        .id("ao-test")
+                        .banner(Banner.builder().format(singletonList(Format.builder().w(300).h(250).build()))
+                                .id("banner_id").build()).ext(mapper.valueToTree(ExtPrebid.of(null,
+                                ExtImpAdocean.of("myao.adocean.pl", "tmYF.DMl7ZBq.Nqt2Bq4FutQTJfTpxCOmtNPZoQUDcL.G7",
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java:139]: Same as above

File: src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+
+    @Test
+    public void makeHttpRequestsShouldSetExpectedHeadersIfDeviceIpIsPresent() {
+        // given
+        final BidRequest bidRequest = BidRequest.builder()
+                .user(User.builder()
+                        .ext(ExtUser.builder()
+                                .consent("COwK6gaOwK6gaFmAAAENAPCAAAAAAAAAAAAAAAAAAAAA.IFoEUQQgAIQwgIwQABAEAAAAOIAAC"
+                                        + "AIAAAAQAIAgEAACEAAAAAgAQBAAAAAAAGBAAgAAAAAAAFAAECAAAgAAQARAEQAAAAAJAAIAAg"
+                                        + "AAAYQEAAAQmAgBC3ZAYzUw").build())
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java:162]: Same as above

File: src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+                                        + "IAAAAQAIAgEAACEAAAAAgAQBAAAAAAAGBAAgAAAAAAAFAAECAAAgAAQARAEQAAAAAJAAIAAgAA"
+                                        + "AYQEAAAQmAgBC3ZAYzUw").build())
+                        .build())
+                .imp(singletonList(Imp.builder()
+                        .id("ao-test")
+                        .banner(Banner.builder().format(singletonList(Format.builder().w(300).h(250).build()))
+                                .id("banner_id").build()).ext(mapper.valueToTree(ExtPrebid.of(null,
+                                ExtImpAdocean.of("myao.adocean.pl", "tmYF.DMl7ZBq.Nqt2Bq4FutQTJfTpxCOmtNPZoQUDcL.G7",
+                                        "adoceanmyaozpniqismex")))).build()))
+                .test(1)
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java:204]: Same as above

File: src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+            final String masterId = params != null ? params.stream()
+                    .filter(param -> param.getName().equals("id"))
+                    .findFirst()
+                    .map(NameValuePair::getValue)
+                    .orElse(null) : null;
+
+            if (masterId != null && masterId.equals(extImpAdocean.getMasterId())) {
+                final String newSlaveId = params.stream()
+                        .filter(param -> param.getName().equals("aid"))
+                        .map(param -> param.getValue().split(":")[0])
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java:128]: ":" - is illegal parameter for URL (encode it)

Or use URLBuilder

File: src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+    @Test
+    public void makeBidsShouldReturnCorrectBidderBid() throws JsonProcessingException {
+        // given
+        final BidRequest bidRequest = BidRequest.builder()
+                .imp(singletonList(Imp.builder()
+                        .id("impId")
+                        .build()))
+                .build();
+
+        final List<AdoceanResponseAdUnit> adoceanResponseAdUnit = asList(AdoceanResponseAdUnit.builder()
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java:287]: \n

File: src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+                        .crid("0af345b42983cc4bc0")
+                        .w(300)
+                        .h(250)
+                        .build(),
+                BidType.banner, "EUR");
+        assertThat(result.getValue().get(0).getBid().getAdm()).isEqualTo(adm);
+        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getValue()).doesNotContainNull().hasSize(1).element(0).isEqualTo(expected);
+    }
+
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java:349]: Add test on .error("true")

File: src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+                }
+                queryParams.add(new BasicNameValuePair("aid", extImpAdocean.getSlaveId() + ":" + impid));
+            }
+        }
+        return false;
+    }
+
+    private String buildUrl(String impid, ExtImpAdocean extImpAdocean, String consentString, Integer test, User user) {
+        final String url = endpointUrl.replace("{{Host}}", extImpAdocean.getEmitterDomain());
+        int randomizedPart = 10000000 + (int) (Math.random() * (99999999 - 10000000));
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java:165]: 99999999 - 10000000 = 89999999

File: src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+        if (test != null && test == 1) {
+            randomizedPart = 10000000;
+        }
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java:168]: you car replace this using ? :

File: src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java [hunk 12]
```diff
@@ -88,56 +89,19 @@ public void makeHttpRequestsShouldReturnErrorIfImpExtCouldNotBeParsed() {
...
-                                        "adoceanmyaozpniqis")))).build()))
-                .test(1)
-
-                .build();
-
-        // when
-        final Result<List<HttpRequest<Void>>> result = adoceanBidder.makeHttpRequests(bidRequest);
-
-        // then
-        assertThat(result.getErrors()).hasSize(1);
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java]: This test is still cover use case. Just rename it to 
makeHttpRequestsShouldSetReturnErrorIfRequestIsDuplicate

File: src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java [hunk 13]
```diff
@@ -347,6 +308,76 @@ public void makeBidsShouldReturnCorrectBidderBid() throws JsonProcessingExceptio
...
+        final BidderBid expected = BidderBid.of(
+                Bid.builder()
+                        .id("ad")
+                        .impid("ao-test")
+                        .adm(adm)
+                        .price(BigDecimal.valueOf(1))
+                        .crid("0af345b42983cc4bc0")
+                        .w(300)
+                        .h(250)
+                        .build(),
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java:374]: Redundant. 
Pls check variables usage. If they are grey in IDE they are not used

File: src/test/resources/org/prebid/server/it/openrtb2/adocean/test-auction-adocean-response.json [hunk 14]
```diff
@@ -7,7 +7,7 @@
           "id": "adoceanmyaozpniqismex",
           "impid": "impId12",
           "price": 10,
-          "adm": " <script> +function() { var wu = \"https://win-url.com\"; var su = \"https://stats-url.com\".replace(/\\[TIMESTAMP\\]/, Date.now()); if (wu && !(navigator.sendBeacon && navigator.sendBeacon(wu))) { (new Image(1,1)).src = wu } if (su && !(navigator.sendBeacon && navigator.sendBeacon(su))) { (new Image(1,1)).src = su } }(); </script>  <!-- code 1 --> ",
+          "adm": "<script>\n\t\t+function() {\n\t\t\tvar wu = \"https://win-url.com\";\n\t\t\tvar su = \"https://stats-url.com\".replace(/\\[TIMESTAMP\\]/, Date.now());\n\n\t\t\tif (wu && !(navigator.sendBeacon && navigator.sendBeacon(wu))) {\n\t\t\t\t(new Image(1,1)).src = wu\n\t\t\t}\n\n\t\t\tif (su && !(navigator.sendBeacon && navigator.sendBeacon(su))) {\n\t\t\t\t(new Image(1,1)).src = su\n\t\t\t}\n\t\t}();\n\t</script> <!-- code 1 --> ",
```
Comment (DGarbar) [inline src/test/resources/org/prebid/server/it/openrtb2/adocean/test-auction-adocean-response.json:10]: don't you think taht \n \t is invalid js ?

**Review by DGarbar — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+            + "\n"
+            + "\t\t\tif (wu && !(navigator.sendBeacon && navigator.sendBeacon(wu))) {\n"
+            + "\t\t\t\t(new Image(1,1)).src = wu\n"
+            + "\t\t\t}\n"
+            + "\n"
+            + "\t\t\tif (su && !(navigator.sendBeacon && navigator.sendBeacon(su))) {\n"
+            + "\t\t\t\t(new Image(1,1)).src = su\n"
+            + "\t\t\t}\n"
+            + "\t\t}();\n"
+            + "\t</script>";
```
[the remark this answers is not in the document]
Comment (AndriyPavlyuk) [reply src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java:66]: This is Java. When we do \n we create separate string. You can look such case in SharethroughMarkupUtil

File: src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+        if (CollectionUtils.isEmpty(request.getImp())) {
+            return Result.emptyWithError(BidderError.badInput("No impression in the bid request"));
+        }
+
+        final User user = request.getUser();
+        final ExtUser extUser = user != null ? user.getExt() : null;
+        final String consent = extUser != null ? extUser.getConsent() : null;
+        final String consentString = StringUtils.isNotBlank(consent) ? consent : "";
+
+        final List<BidderError> errors = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (AndriyPavlyuk) [reply src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java:87]: Why? In Go they collect errors

File: src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java (hunk 12 above)
Comment (AndriyPavlyuk) [reply src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java]: You commented that we do not need to throw exception. In Go they do not add error in this case

File: src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (DGarbar) [reply src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java:87]: Bc you don't collecting them. This collection will have only 1 value

File: src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (DGarbar) [reply src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java:66]: So you are saying that this is perfectly valid JS script ? Pls check Go test case. They removing all the /n /t and set only ine space between code.

**Review by DGarbar — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java [hunk 17]
```diff
@@ -50,20 +50,10 @@
...
-            + "\t\t\tif (wu && !(navigator.sendBeacon && navigator.sendBeacon(wu))) {\n"
-            + "\t\t\t\t(new Image(1,1)).src = wu\n"
-            + "\t\t\t}\n"
-            + "\n"
-            + "\t\t\tif (su && !(navigator.sendBeacon && navigator.sendBeacon(su))) {\n"
-            + "\t\t\t\t(new Image(1,1)).src = su\n"
-            + "\t\t\t}\n"
-            + "\t\t}();\n"
-            + "\t</script>";
+    private static final String MEASUREMENT_CODE = " <script> +function() { var wu = \"%s\";"
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java:53]: Pls change this name to MEASUREMENT_CODE_TEMPLATE

File: src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java [hunk 18]
```diff
@@ -120,53 +98,60 @@ private ExtImpAdocean parseImpExt(Imp imp) {
...
-
-            if (masterId != null && masterId.equals(extImpAdocean.getMasterId())) {
-                final String newSlaveId = queryParams.stream()
-                        .filter(param -> param.getName().equals("aid"))
-                        .map(param -> param.getValue().split(":")[0])
-                        .filter(slaveId -> slaveId.equals(extImpAdocean.getSlaveId()))
+                URIBuilder uriBuilder = new URIBuilder(request.getUri());
+                final List<NameValuePair> queryParams = uriBuilder.getQueryParams();
+
+                final String masterId = queryParams != null
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java:109]: queryParams cant be null

File: src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java [hunk 19]
```diff
@@ -277,20 +277,7 @@ public void makeBidsShouldReturnCorrectBidderBid() throws JsonProcessingExceptio
...
-                + "\t\t\tif (wu && !(navigator.sendBeacon && navigator.sendBeacon(wu))) {\n"
-                + "\t\t\t\t(new Image(1,1)).src = wu\n"
-                + "\t\t\t}\n"
-                + "\n"
-                + "\t\t\tif (su && !(navigator.sendBeacon && navigator.sendBeacon(su))) {\n"
-                + "\t\t\t\t(new Image(1,1)).src = su\n"
-                + "\t\t\t}\n"
-                + "\t\t}();\n"
-                + "\t</script> <!-- code 1 --> ";
+        final String adm = " <script> +function() { var wu = \"https://win-url.com\"; var su = \"https://stats-url.com\".replace(/\\[TIMESTAMP\\]/, Date.now()); if (wu && !(navigator.sendBeacon && navigator.sendBeacon(wu))) { (new Image(1,1)).src = wu } if (su && !(navigator.sendBeacon && navigator.sendBeacon(su))) { (new Image(1,1)).src = su } }(); </script>  <!-- code 1 --> ";
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java:280]: you can refactor this like this 
```
        // then
        final String adm = " <script> +function() { "
                + "var wu = \"https://win-url.com\"; "
                + "var su = \"https://stats-url.com\".replace(/\\[TIMESTAMP\\]/, Date.now()); "
                + "if (wu && !(navigator.sendBeacon && navigator.sendBeacon(wu))) { (new Image(1,1)).src = wu } "
                + "if (su && !(navigator.sendBeacon && navigator.sendBeacon(su))) { (new Image(1,1)).src = su } }(); "
                + "</script>  <!-- code 1 --> ";
```

File: src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java [hunk 20]
```diff
@@ -361,19 +348,6 @@ public void makeBidsShouldReturnEmptyListOfBids() throws JsonProcessingException
                 + "\t\t\t}\n"
                 + "\t\t}();\n"
                 + "\t</script> <!-- code 1 --> ";
-
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/adocean/AdoceanBidderTest.java]: adm redundant

File: src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java [hunk 21]
```diff
@@ -50,20 +50,10 @@
...
-            + "\n"
-            + "\t\t\tif (su && !(navigator.sendBeacon && navigator.sendBeacon(su))) {\n"
-            + "\t\t\t\t(new Image(1,1)).src = su\n"
-            + "\t\t\t}\n"
-            + "\t\t}();\n"
-            + "\t</script>";
+    private static final String MEASUREMENT_CODE = " <script> +function() { var wu = \"%s\";"
+            + " var su = \"%s\".replace(/\\[TIMESTAMP\\]/, Date.now()); if (wu && !(navigator.sendBeacon"
+            + " && navigator.sendBeacon(wu))) { (new Image(1,1)).src = wu } if (su && !(navigator.sendBeacon"
+            + " && navigator.sendBeacon(su))) { (new Image(1,1)).src = su } }(); </script> ";
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/adocean/AdoceanBidder.java:56]:
```
  private static final String MEASUREMENT_CODE = " <script> +function() { " 
            + "var wu = \"%s\"; "
            + "var su = \"%s\".replace(/\\[TIMESTAMP\\]/, Date.now()); " 
            + "if (wu && !(navigator.sendBeacon && navigator.sendBeacon(wu))) { (new Image(1,1)).src = wu } " 
            + "if (su && !(navigator.sendBeacon && navigator.sendBeacon(su))) { (new Image(1,1)).src = su } }(); " 
            + "</script> ";
```

**Review by DGarbar — CHANGES_REQUESTED** [review]

**Review by DGarbar — APPROVED** [review]

**Review by rpanchyk — APPROVED** [review]

**Pull request merged**